[TypeScript & Module Typing] 꼬재(고재증) 미션 제출합니다.

kkojae91 opened the pull request:
## 🔥 팀 미션 적용 결과

본 미션을 수행하며 경험, 영감, 학습한 내용을 팀 미션에 전파해주세요.

- [x] 실제 코드 반영 후 팀 프로젝트에서 타입 테스트 진행
- [x] 자주 사용되는 타입을 유틸로 출시하여 분리

프로젝트 내부에서 PropsWithChildren을 사용을 했었습니다. 하지만 PropsWithChildren의 children이 옵셔널로 잡혀있는 것을 보고 children을 props로 받는다 옵셔널일 필요는 없다고 생각이 들어 PropsWithRecatNodeChildren이라는 유틸 타입을 정의해 사용하고자 합니다.
또한, 저희 프로젝트 구현부에서 중복된 로직을 공통된 컴포넌트로 분리해 사용하기 위해 function을 children으로 넘겨주는 로직이 존재해 PropsWithReactFunctionChildren이라는 유틸 타입을 정의해 사용하고자 합니다.

### [배포된 유틸타입 npm 주소](https://www.npmjs.com/package/pickpick-utils-type)

배포된 유틸타입을 프로젝트에 사용할지에 대해서는 페어와 조금 더 논의후 진행하기로 했습니다.

```ts
export type FunctionChildren<T = any> = (...args: T[]) => ReactNode;

export type PropsWithReactNodeChildren<P = unknown> = P & {
  children: ReactNode;
};

export type PropsWithReactFunctionChildren<
  P = unknown,
  F extends FunctionChildren = (...args: any) => ReactNode
> = P & {
  children: F;
};
```

## ✅ 요구사항 구현 확인

> 필히 테스트 & 타입이 포함

### DOM 유틸

- [x] innerHTML()
- [x] show()
- [x] hidden()
- [x] addEvent()

### 유틸

- [x] fetch()
- [x] isNull()
- [x] isNil()
- [x] isNumber()
- [x] isFunction()
- [x] shuffle()
- [x] pick()
- [x] omit()
- [x] memoize()
- [x] debounce()
- [x] throttle()
- [x] clickOutside()

## 🧐 공유

### 팀 타입스크립트 컨벤션
---

#### 1. 타입스크립트 컴파일 / 빌드 전략

- 로컬과 배포 환경 모두 ts-loader를 사용해 typescript를 컴파일 하고 있다.
- 이유: 현재 프로젝트 규모에서는 babel loader와 ts loader의 빌드 속도 차이가 드라마틱 하지 않기 때문입니다. 추후 프로젝트 규모가 커짐에 따라 빌드 속도가 채감할 정도로 느려져 빌드 타임 때문에 개발 속도에 영향을 끼칠 경우 빌드 속도가 ts-loader보다 빠른 babel-loader 혹은 es build loader를 사용하고, 실제 배포하기 전에는 발견되지 않은 휴먼 에러 방지와 보다 안전한 build를 하기 위해 ts-loader를 사용해 컴파일을 할 계획입니다.


#### 2. type vs interface
- 컴포넌트의 props의 type을 정의할 때는 interface를 쓴다.
- 이유: 컴포넌트의 props 자체는 객체이고 interface로 type을 선언해줄 경우 객체로만 타입을 설계할 수 있기 때문입니다.
- 범용적으로 사용되는 type을 따로 관리 해줘야하는 경우 type alias를 쓴다.
- 이유: interface를 사용할 경우 오버로딩 되는 문제점이 있기 때문에 휴먼 에러가 생길 경우 정의 하지 않는 타입까지 오버로딩되 의도하지 않는 타입 설계가 될 수 있기 때문입니다.


#### 3. 일반 함수 vs 화살표 함수
- 컴포넌트, hook은 일반함수로 작성해 사용한다.
- 이유: 일반 함수는 type 재사용이 불가능하고, 함수 재선언이 가능하다는 단점이 존재하지만, 컴포넌트와 hook의 함수명을 같은 파일 내에서 중복되서 사용할 일이 없고, type 또한 해당하는 컴포넌트와 hook에 맞게 작성하기 때문에 재사용할 일이 없다고 생각했기 때문입니다.
- 그 외의 함수는 화살표함수로 작성해 사용한다.
- 이유: 컴포넌트, hook과 util성 함수를 컨밴션으로 명확히 구분할 수 있어 보다 읽기 좋은 코드를 작성할 수 있다고 생각했습니다. 또한 type 재사용이 가능하고, 중복된 함수명을 사용할 경우 에러가 발생해 프로젝트 규모가 커질 경우 휴먼 에러를 방지할 수 있다고 생각했기 때문입니다.

kkojae91 committed:
chore: 기본 환경 설정

- .gitignore, babel, tsconfig, package.json 설정

kkojae91 committed:
test: _module 내보내지는지 확인하는 테스트 코드 및 _ module 안의 함수들이 잘 포함되어 있는지 확인하는 테스트 코드 작성

kkojae91 committed:
test: _ module 안의 함수들이 잘 동작하는지 확인하는 테스트 코드 작성

- isNull, isNil, isNumber, isFunction 함수 동작 테스트 작성

kkojae91 committed:
chore: tsconfig 불필요한 주석 제거

kkojae91 committed:
test: api test 작성

kkojae91 committed:
test: type test 작성

kkojae91 committed:
feat: util 함수 및 class 메서드 구현

yunjin-kim commented on babel.config.json:
```diff
@@ -0,0 +1,15 @@
+{
```
babel을 쓰신 이유가 있나요? jest config에서 하는 것이 더 좋지 않을까요?

kkojae91 replied:
특별한 이유가 존재하지 않고 터미널 창에 뜬 에러에서 babel에 preset을 적용하면 된다고 나와있어서 babel을 적용했습니다. 😊

babel config 제거 후 jest config 적용했습니다.

yunjin-kim commented on src/index.ts:
```diff
@@ -1,45 +1,201 @@
+    if (Array.isArray(obj)) {
+      const pickArray: T[] = [];
+
+      rest.forEach((restItem) => {
```
forEach말고 다른 내장 메서드를 사용하는 것은 어떤가요? forEach 내부에 로직이 복잡하니 코드를 알아보기 힘드네요 흐규😭흐규😭

kkojae91 replied:
내부로직은 신경쓰지 않고... 작성했습니다! 😊

yunjin-kim commented on src/index.ts:
```diff
@@ -1,45 +1,201 @@
 
-	export function isFunction() {}
+  export function debounce(callback: () => void, delay: number): () => void {
+    let timerId: NodeJS.Timeout;
```
유틸타입인 `ReturnType`을 활용해보는 것은 어떨까요?

kkojae91 replied:
ReturnType 추가로 Parameters에 대해 학습할 수 있었습니다. 감사합니다!

yunjin-kim commented on src/index.ts:
```diff
@@ -1,45 +1,201 @@
+    return omitObject;
+  }
+
+  export function memoize<T>(func: () => T): () => T {
```
인자 타입와 리턴 타입을 살리기 위해 제네릭을 활용해주세욤

kkojae91 replied:
제가 구현한 memoize 함수는 useMemo와 같이 들어온 함수의 리턴값을 메모 해주기 때문에 인자가 존재하지 않아 구현하지 않았어요!

추후 함수 자체를 memoize 하는 함수를 만들 경우 참고해서 작성해보겠습니다~

yunjin-kim reviewed (changes requested):
고생하셨어요

![IMG_4401](https://user-images.githubusercontent.com/79692272/192712914-b563c7a8-ea0f-40c9-a23d-27b61981d7a8.JPG)

많이 지쳐보이던데 힘내세요 핫

kkojae91 committed:
refactor: type 재설정

- is 사용자 정의 타입 가드를 통해 타입 네로잉
- debounce, throttle callback 함수 인자 받을 수 있도록 type 재설정

kkojae91 committed:
chore: babel config 제거 -> jest config 설정
